### PR TITLE
Fix silent data corruption in triu_/tril_ for expanded tensors

### LINKED ISSUE #165987: `triu_/tril_` ignores memory overlaps which can lead to silent correctness issues
### 🐛 Describe the bug

See the sample code below:

```python
In [2]: import torch                                                                                                                                                                        
                                                                                                                                                                                            
In [3]: x = torch.rand((), device="cuda")                                                                                                                                                   
                                                                                                                                                                                            
In [4]: x = x.expand(3, 3)                                                                                                                                                                  
                                                                                                                                                                                            
In [5]: x                                                                                                                                                                                   
Out[5]:                                                                                                                                                                                     
tensor([[0.3047, 0.3047, 0.3047],
        [0.3047, 0.3047, 0.3047],
        [0.3047, 0.3047, 0.3047]], device='cuda:0')

In [6]: x.triu_(1)
Out[6]: 
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]], device='cuda:0')
```
No warning, no error.

### Versions

current main

cc @malfet @jamesr66a

### PR DESCRIPTION
**Pull request by radeksm**
Data corruption occurs in `triu_` and `tril_` if the 2D matrix dimensions contain internal memory overlap (e.g. expanding a 0D scalar to a 2D matrix and calling an in-place triangular operation).

By utilizing `at::assert_no_internal_overlap` we avoid false positives on safely broadcasted batch dimensions, the overlap check is applied exclusively to a 2D of a single matrix from the input after batch dimmension was dropped.

This change prevent data corruption and throws **RuntimeError** with message:

```
  unsupported operation: more than one element of the written-to tensor
  refers to a single memory location. Please clone() the tensor before
  performing the operation.
```

Added positive and negative test scenarios for both `triu_` and `tril_`.

Fixes: #165987

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7407,12 +7407,23 @@ def sample_inputs_tril_triu(op_info, device, dtype, requires_grad, **kwargs):
...
     yield ErrorInput(SampleInput(make_arg((4,))), error_regex="input tensor must have at least 2 dimensions")
 
+    # Cover possible internal memory overlap when e.g. expanding
+    # a scalar 0D to a 2D matrix.
+    x = make_arg(()).expand(3, 3)
+    yield ErrorInput(
+        SampleInput(x, kwargs={'out': x}),
+        error_type=RuntimeError,
+        error_regex="unsupported operation: more than one element of the written-to tensor refers to a single memory location"
+    )
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py]: Not enough --> need to check that `(*, m, 1)` and `(*, 1, n)` do not trigger false positives. Need to test empty tensors, row/col broadcasting, etc.

Comment (radeksm) [reply torch/testing/_internal/common_methods_invocations.py]: > Not enough --> need to check that `(*, m, 1)` and `(*, 1, n)` do not trigger false positives. Need to test empty tensors, row/col

Added few more tests to cover these corner cases.

Although, these new tests seem to be picked up by autograd,  JIT and other tests and now these tests are failing on new triu_/tril_ OpInfos. So put this PR as work in progress again, need to figure out how to make both things working.

Comment (radeksm) [reply torch/testing/_internal/common_methods_invocations.py]: It took me some time to make these new OpInfo working with existing autograd and JIT tests.
@nikitaved  I would appreciate some feedback on these tests please.

Some of the new OpInfo tests only run for forward testing, this prevents backward gradient related tests from failing on expanded or empty tensors.

**Comment by radeksm** [comment]
I ad to remove:
```python
SampleInput(make_tensor((3, 3), dtype=dtype, device=device, requires_grad=False).expand(5, 3, 3))
```

```(3, 3).expand(5, 3, 3)``` tensor is safe for regular tril/triu, but breaks vmap. The problem is that this sample input shouldn't be fed to tests for in-place behavior under vmap.
